### PR TITLE
#2518 -  Disability status minimal solution

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Disability status minimal solution

- [x] Revise the wording on student profile button
image

- [x] Revise the wording on modal header and content of the `Apply for disability status` Modal
image

- [x] Revise the content on student profile disability status ‘tooltip’
image

- [x] Revise the PD/PPD application question and banner content

image

- [x] Created UI level transformation to convert the `DisabilityStatus` enum to following values

```
Not Requested 
Requested 
Approved for Permanent Disability
Approved for Persistent or Prolonged Disability
```

- [x] Created a button in Ministry view of `StudentProfile` to update disability status
image

- [x] Create a modal for ministry update of disability status
image

- [x] Create ministry role `student-update-disability-status` and assign that to `aest-business-administrators` group

### CODE REVIEW

File: sources/packages/forms/src/form-definitions/sfaa2021-22.json [hunk 1]
```diff
@@ -10595,22 +10594,22 @@
               "input": true,
               "tableView": true,
               "inputType": "radio",
-              "label": "<strong>Are you a student with disabilities and wish to be assessed for disability funding?</strong>",
+              "label": "Are you a student who has verified your Disability Status with StudentAid BC as a Permanent Disability (PD) or a Persistent or Prolonged Disability (PPD)?<br><span style=\"font-weight: normal\"> If so, choose option 1.<br> If you have not verified your Disability Status as PD or PPD before, and wish to receive other funding types, choose option 2, complete and submit an Appendix 8 form to StudentAid BC.This may allow you to receive some funding while StudentAid BC verifies your Disability Status. <br> You can find PD and PPD definitions, available disability type funding, eligibility requirements and how to submit your form on <a href=\"https://studentaidbc.ca/sites/all/files/form-library/appendix_8.pdf\" target=\"_blank\" class=\"formio-href\" rel=\"noopener noreferrer\">Appendix 8.</a></span>",
```
Comment (andrewsignori-aot) [inline sources/packages/forms/src/form-definitions/sfaa2021-22.json:10597]: There is a missing white space between "BC.This".

File: sources/packages/web/src/services/http/dto/Student.dto.ts [hunk 2]
```diff
@@ -167,3 +167,10 @@ export interface UpdateSINValidationAPIInDTO {
   expiryDate: string;
   noteDescription: string;
 }
+/**
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/http/dto/Student.dto.ts:170]: Please follow the above methods and add a blank line before the comment.

File: sources/packages/web/src/types/contracts/StudentContract.ts [hunk 3]
```diff
@@ -79,12 +79,27 @@ export enum StudentProfileFormModes {
...
     identityProvider?: IdentityProviders;
     sinConsent: boolean;
+    disabilityStatus: string;
   };
+
+/**
+ * Disability status of student
+ * which is displayed in front end.
+ * Equivalent to {@link DisabilityStatus}
+ *
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/types/contracts/StudentContract.ts:97]: Please remove the blank line.

File: sources/packages/web/src/types/contracts/StudentContract.ts [hunk 4]
```diff
@@ -79,12 +79,27 @@ export enum StudentProfileFormModes {
 
 export type StudentProfileFormModel = Pick<
   StudentProfileAPIOutDTO,
-  "firstName" | "lastName" | "gender" | "email" | "disabilityStatus"
+  "firstName" | "lastName" | "gender" | "email"
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/types/contracts/StudentContract.ts:82]: The removed `disabilityStatus`(boolean) was also a form key in `studentprofile.json` with some logic around it (`"when": "disabilityStatus"`). Should it be removed or adjusted?

File: sources/packages/web/src/components/aest/students/modals/UpdateDisabilityStatusModal.vue [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    <modal-dialog-base
+      title="Update disability status"
+      :showDialog="showDialog"
+    >
+      <template #content>
+        <error-summary :errors="updateDisabilityStatusForm.errors" />
+        <p class="label-value">
+          Update the disability status of student and add note.
+        </p>
+        <v-select
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/modals/UpdateDisabilityStatusModal.vue:12]: We missed many places but I believe that the idea was to use the `hide-details="auto"` to save some vertical spacing. Can we add it? It is the same used for the `v-textarea`.

File: sources/packages/web/src/components/aest/students/DisabilityStatusUpdateTileValue.vue [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      const updatedDisabilityStatus =
+        await updateDisabilityStatusModal.value.showModal();
+      if (updatedDisabilityStatus) {
+        try {
+          await StudentService.shared.updateDisabilityStatus(
+            props.studentId,
+            updatedDisabilityStatus,
+          );
+          snackBar.success("Disability status updated successfully.");
+          context.emit("disabilityStatusUpdated", updatedDisabilityStatus);
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/DisabilityStatusUpdateTileValue.vue:69]: Is the `updatedDisabilityStatus` needed here since the event will call the `StudentService.shared.getStudentProfile` in the consumer?
I am not trying to justify the event args based on the consumer but if there is no need right now I guess that we do not need to provide an event parameter.

File: sources/packages/web/src/components/common/students/StudentProfile.vue [hunk 7]
```diff
@@ -109,27 +113,39 @@ interface SharedStudentProfile extends Omit<StudentProfile, "sin"> {
...
+      required: false,
+      default: false,
+    },
   },
   setup(props) {
     const studentDetail = ref({} as SharedStudentProfile);
     const address = ref({} as AddressAPIOutDTO);
-    const { sinDisplayFormat, emptyStringFiller } = useFormatters();
-    onMounted(async () => {
+    const { sinDisplayFormat, emptyStringFiller, disabilityStatusToDisplay } =
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/common/students/StudentProfile.vue:131]: Is the `disabilityStatusToDisplay` still needed in this component?
I believe that the component `disability-status-update-tile-value` is doing the translation also, right?

File: sources/packages/web/src/components/students/modals/PDStatusApplicationModal.vue [hunk 8]
```diff
@@ -1,20 +1,32 @@
 <template>
   <modal-dialog-base
-    title="Apply for a permanent disability status"
+    title="Apply for a Disability Status"
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/students/modals/PDStatusApplicationModal.vue:3]: I believe that the `modal-dialog-base` can be replaced by the `confirm-modal` with this change.

File: sources/packages/web/src/views/aest/student/Profile.vue [hunk 9]
```diff
@@ -1,6 +1,9 @@
 <template>
   <tab-container>
-    <student-profile :studentId="studentId" />
+    <student-profile
+      :studentId="studentId"
+      :allowDisabilityStatusUpdate="true"
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/Profile.vue:5]: Should this not be defined based on the roles of the Ministry?

File: sources/packages/web/src/components/aest/students/modals/UpdateDisabilityStatusModal.vue [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    const submit = async () => {
+      const validationResult =
+        await updateDisabilityStatusForm.value.validate();
+      if (!validationResult.valid) {
+        return;
+      }
+      // Copying the payload, as reset is making the formModel properties null.
+      const payload = { ...formModel };
+      resolvePromise(payload);
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/components/aest/students/modals/UpdateDisabilityStatusModal.vue:85]: During one of the latest @sh16011993 PR he introduced the `keepModalOpen: true` for the modal `resolvePromise` to allow showing the button loading in the modal while waiting for the API response.
Either we do not have a global loading strategy we have many places where we added the out-of-box button loading at least.
Can you please consider adding the loading to the modal button (following Shashank's changes)?
For this case, the `footer-buttons` has the `processing` to be used.

**Review by andrewsignori-aot — COMMENTED** [review]
Great work, please take a look at the comments.

File: sources/packages/forms/src/form-definitions/sfaa2021-22.json (hunk 1 above)
Comment (ann-aot) [inline sources/packages/forms/src/form-definitions/sfaa2021-22.json:10597]: should this be a part of this qn,
image
 I believe it can be added as a separate HTML component, as we do for other fields, maybe it will come under the options and we can avoid this lengthy qn in the validation,
image
image

File: sources/packages/web/src/types/contracts/StudentContract.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/web/src/types/contracts/StudentContract.ts:82]: the key is not removed, it's data type is modified.

image

File: sources/packages/web/src/components/aest/students/DisabilityStatusUpdateTileValue.vue (hunk 6 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/aest/students/DisabilityStatusUpdateTileValue.vue:69]: Nice catch. Forgot to remove that.

File: sources/packages/web/src/components/students/modals/PDStatusApplicationModal.vue (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/students/modals/PDStatusApplicationModal.vue:3]: I was going towards confirm-modal approach initially, but considering this model to be temporary and going  to go back to it's previous state I took this approach of keeping modal-dialog-base. Let me know your thoughts

File: sources/packages/web/src/views/aest/student/Profile.vue (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/web/src/views/aest/student/Profile.vue:5]: I checked other similar models, and saw that we were using the ministry roles to restrict at only at the primary button of the model and not any earlier in the flow. If I use the ministry role here, it will hide the button to open the modal to all not authorized ministry users unlike the behavior of other models.

**Review by andrepestana-aot — COMMENTED** [review]
Great job. No comment from me now. Please check the other devs' comments.

File: sources/packages/forms/src/form-definitions/sfaa2021-22.json (hunk 1 above)
Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/sfaa2021-22.json:10597]: should this be a part of this qn  - Yes according to the story @ann-aot, but I see your point that this is more a helper text for the question and not the question itself. 

I tried adding that in a separate HTML content and it looked like this(the mandatory asterisks was on all options)

image

And then I went back to the approach what I followed now. 

I will check with @HRAGANBC quickly explaining the impact of having lengthy question vs the HTML approach and see what can be done.

Comment (ann-aot) [reply sources/packages/forms/src/form-definitions/sfaa2021-22.json:10597]: @dheepak-aot, I was thinking about something like this, 
image

File: sources/packages/web/src/components/aest/students/modals/UpdateDisabilityStatusModal.vue (hunk 10 above)
Comment (dheepak-aot) [reply sources/packages/web/src/components/aest/students/modals/UpdateDisabilityStatusModal.vue]: Done.

File: sources/packages/backend/apps/api/src/route-controllers/student/student.aest.controller.ts [hunk 11]
```diff
@@ -389,4 +390,29 @@ export class StudentAESTController extends BaseController {
       throw error;
     }
   }
+
+  /**
+   * Update student disability status with note.
+   * @param studentId student id.
+   * @param payload update disability status payload.
+   */
+  @Roles(Role.StudentUpdateDisabilityStatus)
```
Comment (guru-aot) [inline sources/packages/backend/apps/api/src/route-controllers/student/student.aest.controller.ts:399]: 👍

File: sources/packages/web/src/composables/useFormatters.ts [hunk 12]
```diff
@@ -338,6 +344,25 @@ export function useFormatters() {
...
 
+  /**
+   * Convert the disability status to be displayed
+   * in an equivalent user friendly description.
+   * @param disabilityStatus disability status.
+   * @returns user friendly disability status.
+   */
+  const disabilityStatusToDisplay = (
+    disabilityStatus: DisabilityStatus,
+  ): string => {
```
Comment (ann-aot) [inline sources/packages/web/src/composables/useFormatters.ts:355]: I think the return type can be `DisabilityStatus` right?

File: sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue
Comment (ann-aot) [inline sources/packages/web/src/views/student/financial-aid-application/FullTimeApplication.vue:1]: Not part of the PR, I was wondering why we gave this file name as `fullTime*` 🤔

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @dheepak-aot

File: sources/packages/web/src/components/aest/students/modals/UpdateDisabilityStatusModal.vue [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    </modal-dialog-base>
+  </v-form>
+</template>
+<script lang="ts">
+import { PropType, ref, reactive, defineComponent } from "vue";
+import { useModalDialog, useRules, useFormatters } from "@/composables";
+import { Role, VForm, SelectItemType, DisabilityStatus } from "@/types";
+import { UpdateDisabilityStatusAPIInDTO } from "@/services/http/dto";
+import ErrorSummary from "@/components/generic/ErrorSummary.vue";
+import ModalDialogBase from "@/components/generic/ModalDialogBase.vue";
```
Comment (ann-aot) [inline sources/packages/web/src/components/aest/students/modals/UpdateDisabilityStatusModal.vue:53]: Suggestion: these 2 components are used in many places, may be we can import it to `main.ts`

**Review by ann-aot — COMMENTED** [review]
Nice work @dheepak-aot 👍 Just some minor comments

**Review by andrepestana-aot — APPROVED** [review]
LGTM!

**Review by ann-aot — APPROVED** [review]
Good work @dheepak-aot 👍 Thanks for doing the changes
